src/v/wasm/wasmtime.cc

```diff
@@ -817,13 +826,87 @@ class wasmtime_engine_factory : public factory {
-      , _sr(std::move(sr)) {}
+    explicit wasmtime_runtime(std::unique_ptr<schema_registry> sr)
+      : _sr(std::move(sr)) {
+        wasm_config_t* config = wasm_config_new();
+
+        // Spend more time compiling so that we can have faster code.
+        wasmtime_config_cranelift_opt_level_set(
+          config, WASMTIME_OPT_LEVEL_SPEED);
+        // Fuel allows us to stop execution after some time.
+        wasmtime_config_consume_fuel_set(config, true);
+        // We want to enable memcopy and other efficent memcpy operators
+        wasmtime_config_wasm_bulk_memory_set(config, true);
+        // Our build disables this feature, so we don't need to turn it
+        // off, otherwise we'd want to turn this off (it's on by default).
+        // wasmtime_config_parallel_compilation_set(config, false);
+
+        // Let wasmtime do the stack switching so we can run async host
+        // functions and allow running out of fuel to pause the runtime.
+        //
+        // See the documentation for more information:
+        // https://docs.wasmtime.dev/api/wasmtime/struct.Config.html#asynchronous-wasm
+        wasmtime_config_async_support_set(config, true);
+        // Set max stack size to generally be as big as a contiguous memory
+        // region we're willing to allocate in Redpanda.
+        //
+        // NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
+        wasmtime_config_async_stack_size_set(config, 128_KiB);
+        // The stack size needs to be less than the async stack size, and
+        // whatever is difference between the two is how much host functions can
+        // get, make sure to leave our own functions some room to execute.
+        //
+        // NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
+        wasmtime_config_max_wasm_stack_set(config, 64_KiB);
+        // This disables static memory, see:
+        // https://docs.wasmtime.dev/contributing-architecture.html#linear-memory
+        wasmtime_config_static_memory_maximum_size_set(config, 0_KiB);
+        wasmtime_config_dynamic_memory_guard_size_set(config, 0_KiB);
+        wasmtime_config_dynamic_memory_reserved_for_growth_set(config, 0_KiB);
+        // Don't modify the unwind info as registering these symbols causes C++
+        // exceptions to grab a lock in libgcc and deadlock the Redpanda
+        // process.
+        wasmtime_config_native_unwind_info_set(config, false);
+
+        wasmtime_memory_creator_t memory_creator = {
+          .env = this,
+          .new_memory = &wasmtime_runtime::allocate_heap_memory,
+          .finalizer = nullptr,
+        };
+        wasmtime_config_host_memory_creator_set(config, &memory_creator);
+
+        _engine.reset(wasm_engine_new_with_config(config));
+    }
+
+    ss::future<> start(runtime::config c) override {
+        size_t page_size = ::getpagesize();
+        size_t aligned_pool_size = ss::align_down(
+          c.heap_memory.per_core_pool_size_bytes, page_size);
```

Comment:
if you align_up does the error case go away?

Reply:
No it's still possible for the user to pass in 0.

Reply:
Actually I can add validation for that, and then alignup and not worry about. Still a little sneaky, I don't want to have a divide by zero because someone changes the config later on. I'm going to leave the validation, but update the config.

Reply:
sgtm